In [2]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [18]:
def load_mnist(kind='train'):
    import os
    import gzip
    import numpy as np

    """Load MNIST data from `path`"""
    labels_path = os.path.join(
                               '%s-labels-idx1-ubyte.gz'
                               % kind)
    images_path = os.path.join(
                               '%s-images-idx3-ubyte.gz'
                               % kind)

    with gzip.open(labels_path, 'rb') as lbpath:
        labels = np.frombuffer(lbpath.read(), dtype=np.uint8,
                               offset=8)

    with gzip.open(images_path, 'rb') as imgpath:
        images = np.frombuffer(imgpath.read(), dtype=np.uint8,
                               offset=16).reshape(len(labels), 784)
    return images, labels

In [19]:
X_train, y_train = load_mnist(kind='train')
X_test, y_test = load_mnist(kind='t10k')

In [80]:
from skimage import data, io, filters
from PIL import Image as Image
import numpy as np
import PIL
import matplotlib.pyplot as plt
from tensorflow.keras import losses
from tensorflow.keras import optimizers

In [89]:
np.unique(y_train)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=uint8)

# <center> 1. ReLU

In [91]:
model_relu = Sequential()
model_relu.add(Dense(512, input_shape=(784,), activation="relu"))
model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 39us/sample - loss: nan - accuracy: 0.0988
Epoch 2/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 2s 35us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [================

In [79]:
model_relu = Sequential()
model_relu.add(Dense(128, input_shape=(784,), activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 19us/sample - loss: 3558.9429 - accuracy: 0.1119
Epoch 2/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.3015 - accuracy: 0.1124
Epoch 3/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.3013 - accuracy: 0.1124
Epoch 4/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.3012 - accuracy: 0.1124
Epoch 5/20
60000/60000 [==============================] - 1s 13us/sample - loss: 2.3012 - accuracy: 0.1124
Epoch 6/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.3012 - accuracy: 0.1124
Epoch 7/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.3012 - accuracy: 0.1124
Epoch 8/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.3012 - accuracy: 0.1124
Epoch 9/20
60000/60000 [==============================] - 1s 13us/sample - loss: 2.3012 - accuracy: 0.1124
Epoch 10/20

## ReLU - Tweaks

> ### A. Number of Layers

In [135]:
#512 - 256 - 128 - 64 - 10
model_relu = Sequential()
model_relu.add(Dense(512, input_shape=(784,), activation="relu"))
model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 38us/sample - loss: nan - accuracy: 0.0984
Epoch 2/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 2s 33us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 2s 34us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [================

In [136]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


In [137]:
#256 - 128 - 64 - 10
model_relu = Sequential()
model_relu.add(Dense(256, input_shape=(784,), activation="relu"))
#model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 23us/sample - loss: nan - accuracy: 0.0987
Epoch 2/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 1s 19us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [================

In [138]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


> ### B. Number of nerons in the layers

In [139]:
#256 - 256 - 64 - 10
model_relu = Sequential()
model_relu.add(Dense(256, input_shape=(784,), activation="relu"))
#model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 24us/sample - loss: nan - accuracy: 0.0989
Epoch 2/20
60000/60000 [==============================] - 1s 21us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 1s 20us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [================

In [140]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


In [141]:
#128 - 128 - 64 - 10
model_relu = Sequential()
model_relu.add(Dense(128, input_shape=(784,), activation="relu"))
#model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 24us/sample - loss: nan - accuracy: 0.0988
Epoch 2/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 1s 15us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [================

In [142]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


> ### C. Different batch sizes during training

In [143]:
#256 - 128 - 64 - 10
#batch_size = 8
model_relu = Sequential()
model_relu.add(Dense(256, input_shape=(784,), activation="relu"))
#model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=8, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 10s 159us/sample - loss: nan - accuracy: 0.0987
Epoch 2/20
60000/60000 [==============================] - 9s 156us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 9s 155us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 9s 155us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 9s 155us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 9s 157us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 9s 156us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 9s 156us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 9s 155us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [======

In [144]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


In [145]:
#256 - 128 - 64 - 10
#batch_size = 784
model_relu = Sequential()
model_relu.add(Dense(256, input_shape=(784,), activation="relu"))
#model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=784, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 14us/sample - loss: nan - accuracy: 0.0986         
Epoch 2/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 1s 10us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [=======

In [146]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


In [147]:
#256 - 128 - 64 - 10
#batch_size = 392
model_relu = Sequential()
model_relu.add(Dense(256, input_shape=(784,), activation="relu"))
#model_relu.add(Dense(256, activation="relu"))
model_relu.add(Dense(128, activation="relu"))
model_relu.add(Dense(64, activation="relu"))
model_relu.add(Dense(10, activation="softmax"))

model_relu.compile(optimizer='sgd', loss=losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model_relu.fit(X_train, y_train, batch_size=392, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 16us/sample - loss: nan - accuracy: 0.0987
Epoch 2/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 3/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 4/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 5/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 6/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 7/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 8/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 9/20
60000/60000 [==============================] - 1s 12us/sample - loss: nan - accuracy: 0.0987
Epoch 10/20
60000/60000 [================

In [148]:
score = model_relu.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


<b> Very low accuracy overall

# <center> 2. Tanh

In [98]:
model_tanh = Sequential()
model_tanh.add(Dense(128, input_shape = (784,), activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_38 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_39 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_40 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.8775 - accuracy: 0.7573
Epoch 2/20
60000/60000 [==============================] - 1s 14us/sample - loss: 0.4577 - accuracy: 0.8777
Epoch 3/20
60000/60000 [==============================] - 1s 14us/sample - loss: 0.3836 - accuracy: 0.8928
Epoch 4/20
60000/60000 [=========

## Tanh - Tweaks

> ### A. Number of Layers

In [149]:
#512 - 256 - 128 - 64 - 10
model_tanh = Sequential()
model_tanh.add(Dense(512, input_shape = (784,), activation = 'tanh'))
model_tanh.add(Dense(256, activation = 'tanh'))
model_tanh.add(Dense(128, activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_41"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_166 (Dense)            (None, 512)               401920    
_________________________________________________________________
dense_167 (Dense)            (None, 256)               131328    
_________________________________________________________________
dense_168 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_169 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_170 (Dense)            (None, 10)                650       
Total params: 575,050
Trainable params: 575,050
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 37us/sample - loss: 0.7146 - 

In [150]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.14694015924781562
Test accuracy: 0.9556


In [151]:
#256 - 128 - 64 - 10
model_tanh = Sequential()
model_tanh.add(Dense(256, input_shape = (784,), activation = 'tanh'))
#model_tanh.add(Dense(256, activation = 'tanh'))
model_tanh.add(Dense(128, activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_42"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_171 (Dense)            (None, 256)               200960    
_________________________________________________________________
dense_172 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_173 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_174 (Dense)            (None, 10)                650       
Total params: 242,762
Trainable params: 242,762
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 22us/sample - loss: 0.8412 - accuracy: 0.7782
Epoch 2/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.4183 - accuracy: 0.8886
Epoch 3/

In [152]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16630791448652743
Test accuracy: 0.9499


<b>layers: 512 - 256 - 128 - 64 - 10 performs best

> ### B. Number of nerons in the layers

In [153]:
#256 - 256 - 256 - 64 - 10
model_tanh = Sequential()
model_tanh.add(Dense(256, input_shape = (784,), activation = 'tanh'))
model_tanh.add(Dense(256, activation = 'tanh'))
model_tanh.add(Dense(256, activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_43"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_175 (Dense)            (None, 256)               200960    
_________________________________________________________________
dense_176 (Dense)            (None, 256)               65792     
_________________________________________________________________
dense_177 (Dense)            (None, 256)               65792     
_________________________________________________________________
dense_178 (Dense)            (None, 64)                16448     
_________________________________________________________________
dense_179 (Dense)            (None, 10)                650       
Total params: 349,642
Trainable params: 349,642
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.7342 - 

In [154]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.18029669236540793
Test accuracy: 0.9444


In [155]:
#128 - 128 - 128 - 64 - 10
model_tanh = Sequential()
model_tanh.add(Dense(128, input_shape = (784,), activation = 'tanh'))
model_tanh.add(Dense(128, activation = 'tanh'))
model_tanh.add(Dense(128, activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_44"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_180 (Dense)            (None, 128)               100480    
_________________________________________________________________
dense_181 (Dense)            (None, 128)               16512     
_________________________________________________________________
dense_182 (Dense)            (None, 128)               16512     
_________________________________________________________________
dense_183 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_184 (Dense)            (None, 10)                650       
Total params: 142,410
Trainable params: 142,410
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 21us/sample - loss: 0.8949 - 

In [156]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.21648189881145954
Test accuracy: 0.9337


<b> Best to use narrowing down number of neurons in the layers

> ### C. Different batch sizes during training

In [157]:
#128 - 128 - 64 - 10
model_sig = Sequential()
model_sig.add(Dense(128, input_shape = (784,), activation = 'sigmoid'))
#model_sig.add(Dense(256, activation = 'sigmoid'))
model_sig.add(Dense(128, activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_45"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_185 (Dense)            (None, 128)               100480    
_________________________________________________________________
dense_186 (Dense)            (None, 128)               16512     
_________________________________________________________________
dense_187 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_188 (Dense)            (None, 10)                650       
Total params: 125,898
Trainable params: 125,898
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 25us/sample - loss: 2.2669 - accuracy: 0.2027
Epoch 2/20
60000/60000 [==============================] - 1s 15us/sample - loss: 2.1878 - accuracy: 0.4348
Epoch 3/

In [158]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.21648189881145954
Test accuracy: 0.9337


In [159]:
#512 - 256 - 128 - 64 - 10
#batch_size = full size (784)
model_tanh = Sequential()
model_tanh.add(Dense(512, input_shape = (784,), activation = 'tanh'))
model_tanh.add(Dense(256, activation = 'tanh'))
model_tanh.add(Dense(128, activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 784, epochs = 20, verbose = 1)

Model: "sequential_46"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_189 (Dense)            (None, 512)               401920    
_________________________________________________________________
dense_190 (Dense)            (None, 256)               131328    
_________________________________________________________________
dense_191 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_192 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_193 (Dense)            (None, 10)                650       
Total params: 575,050
Trainable params: 575,050
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 23us/sample - loss: 1.3378 - 

In [160]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.22146317300498486
Test accuracy: 0.9372


In [161]:
#512 - 256 - 128 - 64 - 10
#batch_size = 392
model_tanh = Sequential()
model_tanh.add(Dense(512, input_shape = (784,), activation = 'tanh'))
model_tanh.add(Dense(256, activation = 'tanh'))
model_tanh.add(Dense(128, activation = 'tanh'))
model_tanh.add(Dense(64, activation = 'tanh'))
model_tanh.add(Dense(10, activation = 'softmax'))
model_tanh.summary()

model_tanh.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_tanh.fit(X_train, y_train, batch_size = 392, epochs = 20, verbose = 1)

Model: "sequential_47"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_194 (Dense)            (None, 512)               401920    
_________________________________________________________________
dense_195 (Dense)            (None, 256)               131328    
_________________________________________________________________
dense_196 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_197 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_198 (Dense)            (None, 10)                650       
Total params: 575,050
Trainable params: 575,050
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 26us/sample - loss: 1.0788 - 

In [162]:
score = model_tanh.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.15923606711328028
Test accuracy: 0.9513


# <center> 3. Sigmoid

In [99]:
model_sig = Sequential()
model_sig.add(Dense(128, input_shape = (784,), activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_12"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_41 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_42 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_43 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 17us/sample - loss: 1.9300 - accuracy: 0.5624
Epoch 2/20
60000/60000 [==============================] - 1s 13us/sample - loss: 1.3500 - accuracy: 0.7868
Epoch 3/20
60000/60000 [==============================] - 1s 13us/sample - loss: 0.9790 - accuracy: 0.8329
Epoch 4/20
60000/60000 [=========

## Sigmoid - Tweaks

> ### A. Number of Layers

In [121]:
#512 - 256 - 128 - 64 - 10
model_sig = Sequential()
model_sig.add(Dense(512, input_shape = (784,), activation = 'sigmoid'))
model_sig.add(Dense(256, activation = 'sigmoid'))
model_sig.add(Dense(128, activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_27"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_111 (Dense)            (None, 512)               401920    
_________________________________________________________________
dense_112 (Dense)            (None, 256)               131328    
_________________________________________________________________
dense_113 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_114 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_115 (Dense)            (None, 10)                650       
Total params: 575,050
Trainable params: 575,050
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 38us/sample - loss: 2.3082 - 

In [122]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.5784798972129822
Test accuracy: 0.8702


In [123]:
#256 - 128 - 64 - 10
model_sig = Sequential()
model_sig.add(Dense(256, input_shape = (784,), activation = 'sigmoid'))
#model_sig.add(Dense(256, activation = 'sigmoid'))
model_sig.add(Dense(128, activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_28"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_116 (Dense)            (None, 256)               200960    
_________________________________________________________________
dense_117 (Dense)            (None, 128)               32896     
_________________________________________________________________
dense_118 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_119 (Dense)            (None, 10)                650       
Total params: 242,762
Trainable params: 242,762
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 22us/sample - loss: 2.2600 - accuracy: 0.2526
Epoch 2/20
60000/60000 [==============================] - 1s 18us/sample - loss: 2.1293 - accuracy: 0.4909
Epoch 3/

In [124]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.3172828525543213
Test accuracy: 0.9207


> ### B. Number of nerons in the layers

In [125]:
#256 - 256 - 64 - 10
model_sig = Sequential()
model_sig.add(Dense(256, input_shape = (784,), activation = 'sigmoid'))
#model_sig.add(Dense(256, activation = 'sigmoid'))
model_sig.add(Dense(256, activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_29"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_120 (Dense)            (None, 256)               200960    
_________________________________________________________________
dense_121 (Dense)            (None, 256)               65792     
_________________________________________________________________
dense_122 (Dense)            (None, 64)                16448     
_________________________________________________________________
dense_123 (Dense)            (None, 10)                650       
Total params: 283,850
Trainable params: 283,850
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 24us/sample - loss: 2.2622 - accuracy: 0.2279
Epoch 2/20
60000/60000 [==============================] - 1s 20us/sample - loss: 2.1389 - accuracy: 0.5110
Epoch 3/

In [126]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.3005338268995285
Test accuracy: 0.9222


In [127]:
#128 - 128 - 64 - 10
model_sig = Sequential()
model_sig.add(Dense(128, input_shape = (784,), activation = 'sigmoid'))
#model_sig.add(Dense(256, activation = 'sigmoid'))
model_sig.add(Dense(128, activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 128, epochs = 20, verbose = 1)

Model: "sequential_30"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_124 (Dense)            (None, 128)               100480    
_________________________________________________________________
dense_125 (Dense)            (None, 128)               16512     
_________________________________________________________________
dense_126 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_127 (Dense)            (None, 10)                650       
Total params: 125,898
Trainable params: 125,898
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 24us/sample - loss: 2.2859 - accuracy: 0.1878
Epoch 2/20
60000/60000 [==============================] - 1s 15us/sample - loss: 2.1858 - accuracy: 0.4121
Epoch 3/

In [128]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.3650088099956512
Test accuracy: 0.909


> ### C. Different batch sizes during training

In [129]:
#batch_size = 8
model_sig = Sequential()
model_sig.add(Dense(128, input_shape = (784,), activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 8, epochs = 20, verbose = 1)

Model: "sequential_31"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_128 (Dense)            (None, 128)               100480    
_________________________________________________________________
dense_129 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_130 (Dense)            (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 8s 135us/sample - loss: 0.8156 - accuracy: 0.8118
Epoch 2/20
60000/60000 [==============================] - 8s 132us/sample - loss: 0.4449 - accuracy: 0.8831
Epoch 3/20
60000/60000 [==============================] - 8s 133us/sample - loss: 0.4110 - accuracy: 0.8884
Epoch 4/20
60000/60000 [======

In [130]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.34047010437250136
Test accuracy: 0.8945


In [131]:
#batch_size = 784
model_sig = Sequential()
model_sig.add(Dense(128, input_shape = (784,), activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 784, epochs = 20, verbose = 1)

Model: "sequential_32"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_131 (Dense)            (None, 128)               100480    
_________________________________________________________________
dense_132 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_133 (Dense)            (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 10us/sample - loss: 2.3421 - accuracy: 0.1802
Epoch 2/20
60000/60000 [==============================] - 0s 6us/sample - loss: 2.1168 - accuracy: 0.4039
Epoch 3/20
60000/60000 [==============================] - 0s 7us/sample - loss: 1.9981 - accuracy: 0.5288
Epoch 4/20
60000/60000 [===========

In [132]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.8303745219230652
Test accuracy: 0.859


In [133]:
#batch_size = 392
model_sig = Sequential()
model_sig.add(Dense(128, input_shape = (784,), activation = 'sigmoid'))
model_sig.add(Dense(64, activation = 'sigmoid'))
model_sig.add(Dense(10, activation = 'softmax'))
model_sig.summary()

model_sig.compile(optimizer = 'sgd', loss = losses.sparse_categorical_crossentropy, metrics = ['accuracy'])

model_sig.fit(X_train, y_train, batch_size = 392, epochs = 20, verbose = 1)

Model: "sequential_33"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_134 (Dense)            (None, 128)               100480    
_________________________________________________________________
dense_135 (Dense)            (None, 64)                8256      
_________________________________________________________________
dense_136 (Dense)            (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 12us/sample - loss: 2.1553 - accuracy: 0.3595
Epoch 2/20
60000/60000 [==============================] - 1s 9us/sample - loss: 1.8945 - accuracy: 0.5738
Epoch 3/20
60000/60000 [==============================] - 1s 8us/sample - loss: 1.7005 - accuracy: 0.6553
Epoch 4/20
60000/60000 [===========

In [134]:
score = model_sig.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.4782491657733917
Test accuracy: 0.895


# <center>CONCLUSION

<b>Best Model: Tanh</b><br>

<b>tweaks:</b>
> 5 Layers: 512 - 256 - 128 - 64 - 10 <br>
> Batch Size = 128 <br>
<br>

> Scores:<br>
> Test score: 0.14694015924781562 <br>
> Test accuracy: 0.9556 <br>
